### ISSUE TITLE
Juggling async; learnyounode 9; why won't this code work?

### ISSUE THREAD

**Issue by Holstiche** [nodeschool/discussions#1632]
So this is my code for the juggling async exercise.
```
http = require('http');

data_array = [];
init_count = 0;
done_count = 0;

function output() {
	if (init_count === done_count){
		console.log(data_array[0]);
		console.log(data_array[1]);
		console.log(data_array[2]);
	}
};

function data_to_array1(url, number) {
	init_count = init_count + 1;
	http.get(url, function (response) {
		ret_data = '';
		response.setEncoding('utf8');
		response.on('data', function (data){
			ret_data = ret_data + data;
		});
		response.on('end', function(){
			done_count = done_count + 1;
			data_array[number-1] = ret_data;
			output();
		});
		response.on('error', function (error){
			done_count = done_count + 1;
			console.log(error)
			output();

		});
	} );
};

data_to_array1(process.argv[2], 1);
data_to_array2(process.argv[3], 2);
data_to_array3(process.argv[4], 3);
```

The problem is that sometimes when you run it it works, sometimes only one or two are correct and sometimes all the answers sort of mix. Could it have to do with the fact that they use the same function, so that when the data comes in at the same time they store it in the same variable? I have read a lot of other issues but I could not find the problem anywhere. Thanks in advance!

**Comment by ed-chin** [comment]
There's a few things wrong with this:
1) where did you define the function data_to_array2? 
2) if (init_count === done_count) could be a race condition the way you set it up. Just check if (done_count== 3)
3) your concat might be backwards. "ret_data = ret_data + data;"
4) why aren't you using bufferList or stream concat as were given in the instructions?

**Comment by radiofreejohn** [comment]
@Holstiche I think the problem here is that you're defining `ret_data` without putting `var` ahead of it, making it a global variable. 

`ret_data` is getting re-used between the three different calls, so there will be very weird results unless you make sure to declare it as a local variable to the function like:

`var ret_data = ''`

**Comment by Holstiche** [comment]
@ed-chin Thanks for the response
1) I tried if it would work if I defined three different functions, so that is were the naming came from. I forgot to change them when I used 1 function again.
2) Yeah I noticed that, but I think the chance of a function finishing before the next one has started if very small. I am still thinking about another way to write is, so any suggestions would be really helpfull.
3) I don't think thats true because 'data' contains the new data and 'ret_data' contains the already appended data.
4) I just wanted to try it out this way.

@radiofreejohn You were right, declaring ret_data as a global variable was the problem. Thanks a lot!
